pkg/csi/cinder/controllerserver.go

```diff
@@ -110,31 +112,56 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 	}
 	content := req.GetVolumeContentSource()
 	var snapshotID string
-	var sourcevolID string
+	var sourceVolID string
+	var sourceBackupID string
 
 	if content != nil && content.GetSnapshot() != nil {
 		snapshotID = content.GetSnapshot().GetSnapshotId()
-		_, err := cloud.GetSnapshotByID(snapshotID)
-		if err != nil {
-			if cpoerrors.IsNotFound(err) {
-				return nil, status.Errorf(codes.NotFound, "VolumeContentSource Snapshot %s not found", snapshotID)
+
+		snap, snapErr := cs.Cloud.GetSnapshotByID(snapshotID)
+		if snapErr != nil && !cpoerrors.IsNotFound(snapErr) {
+			return nil, snapErr
+		}
+		// If the snapshot exists but is not yet available, fail.
+		if snapErr == nil && snap.Status != "available" {
+			return nil, status.Errorf(codes.NotFound, "VolumeContentSource Snapshot %s is not yet available. status: %s", snapshotID, snap.Status)
+		}
+
+		// In case a snapshot is not found
+		// check if a Backup with the same ID exists
+		if cpoerrors.IsNotFound(snapErr) {
+			back, err := cs.Cloud.GetBackupByID(snapshotID)
+			if err != nil {
+				//If there is an error getting the backup as well, fail.
+				return nil, status.Errorf(codes.NotFound, "VolumeContentSource Snapshot or Backup with ID %s not found", snapshotID)
+			}
```

Comment:
So, since we have the same interface for looking for an snapshot and looking for a backup, we first check if a snapshot with name N exists, and if it doesn't, we check for a backup with name N. However, some systems will not support cinder backups as they are not required. 

This will lead to this `err` being a "real" (not not found) error many times, but it is not necessarily an issue.

So:

if the system does not support backups, and there is an error: No issue, we should return codes.NotFound due to snapshot not being found
 
if the system supports backups, and this is a NotFound error: No issue, we should return codes.NotFound due to snapshot and backup not being found

if the system supports backups, and this has a real error: Oops!, we should have sent codes.Internal

So we are covering 2/3 cases. Since we can't differentiate cases 1 and 3 (because we can't know if the cinder installation supports backups), we should aim for the case where the users that aren't using this specific feature do not get affected

Reply:
I don't think API will behave differently if `cinder-backup` is not deployed. In such case API will still get us a 404, as `GET` call doesn't RPC down to `cinder-backup` and just looks the items up in the DB: https://opendev.org/openstack/cinder/src/branch/master/cinder/backup/api.py#L71-L76.

If backups are optional, wouldn't it be better to put all the backup operations behind an `if` based on existence of the backup API extension: https://docs.openstack.org/api-ref/block-storage/v3/#list-known-api-extensions ?